Allow to set IP address to analyze

marcstern opened the pull request:
This allows to use another mechanism than mod_remoteip to set the origin IP.
Typically, mod_security2 can be used (in phase 1).
Totally compatible with the current behaviour if this feature is not used.

nchelluri commented:
Hi marcstern, looks good!

Can you please:
1. Add a test for this functionality so that we don't accidentally break it in the future,
2. Add documentation for this, I am happy to proofread/edit it
3. Reword the commit message to "Allow setting IP address to analyze"

Then we can merge this.

marcstern commented:
I modified the documentation.

For the tests, I don't know how to code them. I image I have to modify the config to add a specific location, then to add a Perl script to launch a request to that URL and check the result... No idea exactly how.

Note that I don't know how to test this with standard httpd modules. I tested it using mod_security2 with non-default compilation option to set the variable before mod_maxminddb. Can we include such a module? Or we could create a specific module to implement a "setenv" that runs before mod_maxminddb but both solutions look very heavy.

marcstern commented:
@nchelluri: can I do something else?

nchelluri commented:
Hi @marcstern and sorry for the late reply. I am trying to write a test for this and running into some difficulty injecting the environment variable into the test Apache instance. You definitely tested this against a real Apache instance, right?

nchelluri commented:
Can you give me your modsecurity config to set the environment variable?

marcstern commented:
You need to compile ModSecurity with "configure --enable-request-early". Then:
SecAction "phase:1,nolog,pass,setenv:MMDB_ADDR=8.8.8.8"

marcstern commented:
mod_env runs after this module; this is why it doesn't work.
mod_security2 runs (in phase 1) before this module; this is why it's working.

nchelluri commented:
Hi @marcstern , unfortunately after consulting with my colleagues we've decided against merging this fix. This is because it is quite tough to get the environment variable into `mod_maxminddb` and we feel that this will create more confusion than it will solve.

I think you should keep your 1-line fork around; I believe we are unlikely to break it any time soon. Thanks for your work on this.

marcstern commented:
Actually, there's a very easy way to test it: set the environment variable in the shell before launching httpd.
About the confusion, couldn't we add this line inside a conditional compile with the explanation? We would have the best of both worlds: no change in the doc, no change in the normal builds, and the possibility for the "power developers" that need it.

marcstern commented:
I'll try to setup this.
It'll be for July, not possible before

nchelluri commented:
Excellent!

marcstern commented:
I tried to implement th test by monkeying what exists:
 Added mod_security2 build & install.
 Added directives in config.
 Added a test to trigger it from the URL (query string).
I couldn't test it, but I expect it to be almost correct.
I hope you'll be able to fix some potential syntax problems :-(
Thanks

marcstern commented:
My test install works, but the general Apache setup (from the trunk) is incorrect.
I tried several things but other problems are continuously coming.
By fixing the trunk, my tests should be OK.
